File: atcoder/abc/go/abc219/abc219_H/abc219_H.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"sort"
	"strconv"
)

type scanner struct { sc *bufio.Scanner }
func newScanner(input io.Reader) *scanner {
	sc := bufio.NewScanner(input)
	sc.Split(bufio.ScanWords)
	sc.Buffer(make([]byte, 1024), int(1e+9))
	return &scanner{sc}	
}
var rdr = newScanner(os.Stdin)
const BUFSIZE = 10000000
var wrtr = bufio.NewWriterSize(os.Stdout, BUFSIZE)
func gs() string  { rdr.sc.Scan(); return rdr.sc.Text() }
func gi() int     { i,e := strconv.Atoi(gs()); if e != nil {panic(e)}; return i }
func ia(m int) []int { return make([]int,m) }
func fill2(m int) ([]int,[]int) { a,b := ia(m),ia(m); for i:=0;i<m;i++ {a[i],b[i] = gi(),gi()}; return a,b }
func max(a,b int) int { if a > b { return a }; return b }

type candle struct {x,a int}
func main() {
	//f1, _ := os.Create("cpu.prof"); pprof.StartCPUProfile(f1); defer pprof.StopCPUProfile()
    defer wrtr.Flush()
	infn := ""
	if infn == "" && len(os.Args) > 1 {	infn = os.Args[1] }
    if infn != "" {	f, e := os.Open(infn); if e != nil { panic(e) }; rdr = newScanner(f) }
	N := gi(); X,A := fill2(N)
	lf := make([]candle,0); rt := make([]candle,0)
	for i:=0;i<N;i++ { if X[i] < 0 { lf = append(lf,candle{-X[i],A[i]}) } else { rt = append(rt,candle{X[i],A[i]}) } }
	sort.Slice(lf,func(i,j int)bool{return lf[i].x < lf[j].x})
	sort.Slice(rt,func(i,j int)bool{return rt[i].x < rt[j].x})

	dp := [301][301][2][301]int{}
	maxi,maxj,inf,dist := len(lf),len(rt),1_000_000_000_000_000_000,0
	for i:=maxi;i>=0;i-- {
		for j:=maxj;j>=0;j-- {
			for k:=0;k<2;k++ {
				for l:=0;l<=N;l++ { // Number of candles left
					if l == 0 { dp[i][j][k][l] = 0; continue }
					curx := 0; if k == 0 && i > 0 { curx = lf[i-1].x }; if k == 1 && j > 0 { curx = rt[j-1].x }
					v := -inf
					if i < maxi {
						if k == 0 { dist = lf[i].x - curx } else { dist = lf[i].x + curx }
						v = max(v,-dist*l+lf[i].a+dp[i+1][j][0][l-1])
						v = max(v,-dist*l+dp[i+1][j][0][l])
					}
					if j < maxj {
						if k == 0 { dist = rt[j].x + curx } else { dist = rt[j].x - curx }
						v = max(v,-dist*l+rt[j].a+dp[i][j+1][1][l-1])
						v = max(v,-dist*l+dp[i][j+1][1][l])
					}
					dp[i][j][k][l] = v
				}
			}
		}
	}
	best := 0
	for k:=0;k<2;k++ {
		for l:=0;l<=N;l++ {
			best = max(best,dp[0][0][k][l])
		}
	}
	fmt.Println(best)
}

File: abc/python/abc219/abc219_H.py

import sys
infile = sys.stdin.buffer

def gs()  : return infile.readline().rstrip()
def gi()  : return int(gs())
def gf()  : return float(gs())
def gss() : return gs().split()
def gis() : return [int(x) for x in gss()]
def gfs() : return [float(x) for x in gss()]

def main(infn="") :
    global infile
    infile = open(infn,"r") if infn else open(sys.argv[1],"r") if len(sys.argv) > 1 else sys.stdin
    N = gi()
    X = []; A = []
    for _ in range(N) : x,a = gis(); X.append(x); A.append(a)
    lf = [(0,0)]; rt = [(0,0)]
    for x,a in zip(X,A) : 
        if x < 0 : lf.append((-x,a))
        else     : rt.append((x,a))
    lf.sort(); rt.sort(); nleft = len(lf); nright = len(rt); myinf = 10**18
    dp = [0] * nleft
    for i in range(nleft) : 
        dp[i] = [0] * nright
        for j in range(nright) :
            dp[i][j] = [0] * (N+1)
            for k in range(N+1) :
                dp[i][j][k] = [-myinf] * 2
    for i in range(nleft-1,-1,-1) :
        for j in range(nright-1,-1,-1) :
            for k in range(N+1) :
                for l in range(2) :
                    if k == 0 : dp[i][j][k][l] = 0; continue 
                    v = -myinf
                    if i+1 < nleft :
                        dist = lf[i+1][0] - lf[i][0] if l == 0 else lf[i+1][0] + rt[j][0]
                        v = max(v, -dist*k + lf[i+1][1] + dp[i+1][j][k-1][0], -dist*k + dp[i+1][j][k][0])
                    if j+1 < nright :
                        dist = rt[j+1][0] - rt[j][0] if l == 1 else rt[j+1][0] + lf[i][0]
                        v = max(v, -dist*k + rt[j+1][1] + dp[i][j+1][k-1][1], -dist*k + dp[i][j+1][k][1])
                    dp[i][j][k][l] = v
    best = 0
    for k in range(N+1) :
        for l in range(2) :
            best = max(best,dp[0][0][k][l])
    print(best)

if __name__ == '__main__' :
    main()
    sys.stdout.flush()

